File: Problems/No.1358_Number_of_Substrings_Containing_All_Three_Characters/number_of_substrings_containing_all_three_characters.py
class Solution(object):
    def numberOfSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        """

        # parameters
        count = 0
        sLen = len(s)
        char_count = {'a': 0, 'b': 0, 'c': 0}

        # start computing
        substr_start = 0
        for substr_end, c in enumerate(s):
            char_count[c] = char_count.get(c, 0) + 1
            while ((char_count['a'] > 0) and (char_count['b'] > 0) and (char_count['c'] > 0) and
                   (substr_start <= substr_end)):
                char_count[s[substr_start]] = char_count.get(s[substr_start], 0) - 1
                substr_start += 1
                count += sLen - substr_end

        return count

File: Problems/No.1358_Number_of_Substrings_Containing_All_Three_Characters/number_of_substrings_containing_all_three_characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        
        // parameters
        int count = 0;
        int sLen = s.length();
        std::unordered_map<char, int> char_count = {{'a', 0}, {'b', 0}, {'c', 0}};

        // start computing
        int substr_start = 0;
        for (int substr_end=0; substr_end<sLen; substr_end++) {

            char_count[s[substr_end]]++;

            while ((char_count['a'] > 0) && (char_count['b'] > 0) && (char_count['c'] > 0) && 
                   (substr_start <= substr_end)) {

                char_count[s[substr_start++]]--;
                count += sLen - substr_end;
               
            }

        }

        return count;

    }
};
